src/config.cpp

```diff
@@ -478,5 +479,127 @@ const std::string& CircuitConfig::getExpandedJSON() const {
+{
+  public:
+    Parser(const std::string& content, const std::string& basePath)
+        : _basePath(fs::absolute(fs::path(basePath)).lexically_normal())
+        , _json(nlohmann::json::parse(content)) {}
+
+    template <typename Iterator, typename Type, typename SectionName>
+    void parseMandatory(const Iterator& it,
+                        const char* name,
+                        const SectionName& sn,
+                        Type& buf) const {
+        const auto element = it.find(name);
+        if (element == it.end())
+            throw SonataError(fmt::format("Could not find '{}' in '{}'", name, sn));
+        buf = element->template get<Type>();
+    }
+
+    template <typename Iterator, typename Type>
+    void parseOptional(const Iterator& it, const char* name, Type& buf) const {
+        const auto element = it.find(name);
+        if (element != it.end())
+            buf = element->template get<Type>();
+    }
+
+    SimulationConfig::Run parseRun() const {
+        const auto runIt = _json.find("run");
+        if (runIt == _json.end())
+            throw SonataError("Could not find 'run' section");
+
+        SimulationConfig::Run result{};
+        parseMandatory(*runIt, "tstop", "run", result.tstop);
+        parseMandatory(*runIt, "dt", "run", result.dt);
+        return result;
+    }
+
+    SimulationConfig::Output parseOutput() const {
+        SimulationConfig::Output result{};
+        result.outputDir = "output";
+        result.spikesFile = "out.h5";
+
+        const auto outputIt = _json.find("output");
+        if (outputIt != _json.end()) {
+            parseOptional(*outputIt, "output_dir", result.outputDir);
+            parseOptional(*outputIt, "spikes_file", result.spikesFile);
+        }
+
+        result.outputDir = toAbsolute(_basePath, result.outputDir);
+
+        return result;
+    }
+
+    std::unordered_map<std::string, SimulationConfig::Report> parseReports() const {
+        std::unordered_map<std::string, SimulationConfig::Report> result;
+
+        const auto reportsIt = _json.find("reports");
+        if (reportsIt == _json.end())
+            return result;
+
+        for (auto it = reportsIt->begin(); it != reportsIt->end(); ++it) {
+            auto& report = result[it.key()];
+            auto& valueIt = it.value();
+            const auto debugStr = fmt::format("report {}", it.key());
+            parseMandatory(valueIt, "cells", debugStr, report.cells);
+            parseMandatory(valueIt, "type", debugStr, report.type);
+            parseMandatory(valueIt, "dt", debugStr, report.dt);
+            parseMandatory(valueIt, "start_time", debugStr, report.startTime);
+            parseMandatory(valueIt, "end_time", debugStr, report.endTime);
+            parseOptional(valueIt, "file_name", report.fileName);
+            if (report.fileName.empty())
+                report.fileName = it.key() + "_SONATA.h5";
+            else if (report.fileName.find(".h5") == std::string::npos)
```

Comment:
Looks like this should be checking the string ending, not searching for `.h5` anywhere in it.
(Sadly [std::string::ends_with](https://en.cppreference.com/w/cpp/string/basic_string/ends_with) is C++20...)

Reply:
Good point, there are many weird paths that could make this fail. Ill commit a better version and ping you, thanks a lot!

Reply:
Done!, please have a look.